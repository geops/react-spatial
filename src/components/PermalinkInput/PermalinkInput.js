import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { MdContentCopy } from 'react-icons/md';
import Button from '../Button';

const propTypes = {
  /**
   * Value of the permalink
   */
  value: PropTypes.string,

  /**
   * CSS class of the container (input+button).
   */
  className: PropTypes.string,

  /**
   * CSS class of the input field.
   */
  classNameInputField: PropTypes.string,

  /**
   * CSS class of the copy button.
   */
  classNameCopyBt: PropTypes.string,

  /**
   * Content for copy button.
   */
  button: PropTypes.any,

  /**
   * Title for the copy button.
   */
  titleCopyBt: PropTypes.string,

  /**
   * Title for the input field.
   */
  titleInputField: PropTypes.string,

  /**
   * Callback function to get a shortened URL.
   */
  getShortenedUrl: PropTypes.func,
};

const defaultProps = {
  value: '',
  className: 'tm-permalink-field',
  classNameInputField: 'tm-permalink-input',
  classNameCopyBt: 'tm-permalink-bt',
  titleCopyBt: '',
  titleInputField: '',
  button: <MdContentCopy focusable={false} />,
  getShortenedUrl: val => {
    return new Promise(resolve => {
      return resolve(val);
    });
  },
};

/**
 * This component displays a permalink field.
 */
class PermalinkInput extends PureComponent {
  static selectInput() {
    document.execCommand('selectall');
  }

  constructor(props) {
    super(props);
    this.state = { permalinkValue: null };
    this.inputRef = null;
  }

  componentDidMount() {
    const { value } = this.props;

    if (value) {
      this.updatePermalinkValue();
    }
  }

  componentDidUpdate(prevProps) {
    const { value } = this.props;

    if (value !== prevProps.value) {
      this.updatePermalinkValue();
    }
  }

  onClickCopyButton() {
    if (this.inputRef) {
      this.inputRef.select();
    }
    document.execCommand('copy');
  }

  updatePermalinkValue() {
    const { value, getShortenedUrl } = this.props;

    return getShortenedUrl(value).then(v => {
      this.setState({ permalinkValue: v });
    });
  }

  render() {
    const {
      button,
      className,
      classNameInputField,
      classNameCopyBt,
      titleCopyBt,
      titleInputField,
    } = this.props;

    const { permalinkValue } = this.state;
    return (
      <div className={className}>
        <input
          value={permalinkValue || ''}
          type="text"
          tabIndex="0"
          title={titleInputField}
          className={classNameInputField}
          ref={node => {
            this.inputRef = node;
          }}
          onClick={() => PermalinkInput.selectInput()}
          onChange={() => {}}
        />
        <Button
          className={classNameCopyBt}
          title={titleCopyBt}
          onClick={() => {
            this.onClickCopyButton();
          }}
        >
          {button}
        </Button>
      </div>
    );
  }
}

PermalinkInput.propTypes = propTypes;
PermalinkInput.defaultProps = defaultProps;

export default PermalinkInput;
