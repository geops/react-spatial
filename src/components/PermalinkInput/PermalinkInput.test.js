import React from 'react';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import PermalinkInput from './PermalinkInput';

configure({ adapter: new Adapter() });

describe('PermalinkInput', () => {
  test('matches snapshot', () => {
    const component = renderer.create(
      <PermalinkInput value="http://url.test" />,
    );
    expect(component.toJSON()).toMatchSnapshot();
    component.getInstance().setState({ permalinkValue: 'http://url.test' });
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe('when interacts,', () => {
    let getShortenedUrl = null;

    beforeEach(() => {
      getShortenedUrl = jest.fn(val => {
        return new Promise(resolve => {
          return resolve(val);
        });
      });
    });

    test('Promise resolution set input value.', () => {
      const wrapper = shallow(
        <PermalinkInput getShortenedUrl={getShortenedUrl} />,
      );
      wrapper.setProps({ value: 'http://url.test' });
      wrapper
        .instance()
        .updatePermalinkValue()
        .then(() => {
          const input = wrapper
            .find('input')
            .first()
            .getElement();

          expect(input.props.value).toBe('http://url.test');
        });
    });

    test('getShortenedUrl is called to set value.', () => {
      shallow(
        <PermalinkInput
          getShortenedUrl={getShortenedUrl}
          value="http://url.test"
        />,
      );

      expect(getShortenedUrl).toHaveBeenCalledTimes(1);
      expect(getShortenedUrl).toHaveBeenCalledWith('http://url.test');
    });

    test('select value on input click.', () => {
      document.execCommand = jest.fn();
      const wrapper = mount(
        <PermalinkInput
          getShortenedUrl={getShortenedUrl}
          value="http://url.test"
        />,
      );
      const spy = jest.spyOn(PermalinkInput, 'selectInput');

      expect(spy).toHaveBeenCalledTimes(0);

      wrapper
        .instance()
        .updatePermalinkValue()
        .then(() => {
          wrapper.update();
          wrapper
            .find('input')
            .first()
            .simulate('click');

          expect(spy).toHaveBeenCalledTimes(1);
        });
    });

    test('click button copy the value.', () => {
      document.execCommand = jest.fn();
      const wrapper = mount(
        <PermalinkInput
          getShortenedUrl={getShortenedUrl}
          value="http://url.test"
        />,
      );
      const spy = jest.spyOn(wrapper.instance(), 'onClickCopyButton');

      expect(spy).toHaveBeenCalledTimes(0);

      wrapper
        .instance()
        .updatePermalinkValue()
        .then(() => {
          wrapper.update();
          wrapper
            .find('.tm-permalink-bt')
            .first()
            .simulate('click');
          expect(spy).toHaveBeenCalledTimes(1);
        });
    });
  });
});
