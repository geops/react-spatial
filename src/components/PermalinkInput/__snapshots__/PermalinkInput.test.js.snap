// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PermalinkInput matches snapshot 1`] = `null`;

exports[`PermalinkInput matches snapshot 2`] = `
<div
  className="tm-permalink-field"
>
  <input
    className="tm-permalink-input"
    onClick={[Function]}
    tabIndex="0"
    title=""
    type="text"
    value="http://url.test"
  />
  <div
    aria-label=""
    className="tm-permalink-bt"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
    title=""
  >
    <svg
      fill="currentColor"
      focusable={false}
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm0 16H8V7h11v14z"
      />
    </svg>
  </div>
</div>
`;
